File: leetcode169.cpp
// Divide and Conquer
//时间复杂度 O(nlgn)
class Solution {
private:
	int majority(vector<int>& nums, int left, int right)
	{
		if(left == right) return nums[left];
		int mid = left + ((right - left)>>1);
		int lm = majority(nums,left,mid);
		int rm = majority(nums,mid+1,right);
		if(lm == rm) return lm;
		return return count(nums.begin() + left, nums.begin() + right + 1, lm) > count(nums.begin() + left, nums.begin() + right + 1, rm) ? lm : rm;
	}
public:
    int majorityElement(vector<int>& nums) {
        return majority(nums, 0, nums.size()-1);
    }
};

File: leetcode169.java
//divide and conquer
public class Solution {
    public int majorityElement(int[] nums) {
    	return helper(nums,0,nums.length-1);
    }

    private int helper(int[] nums, int l, int r){
    	if(l == r)
    		return nums[l];
    	int m = (l+r)/2;
    	int a = helper(nums,l,m);
    	int b = helper(nums,m+1,r);

    	if(a == b)
    		return a;
    	int count1 = 0, count2 = 0;
    	for(int i = l; i <= r; i++){
    		if(nums[i] == a)
    			count1++;
    		else if(nums[i] == b)
    			count2++;
    	}
    	return count1 > count2 ? a : b;
    }
}
